include/tc/core/cuda/cuda_mapping_options.h

```diff
@@ -0,0 +1,248 @@
+  inline bool operator==(const CudaDimView& view) const;
+  inline bool operator!=(const CudaDimView& view) const;
+
+  /// Conversion to string and output operators.
+  std::string toCommaSeparatedString() const;
+  friend std::ostream& operator<<(std::ostream& os, const CudaDimView& view);
+
+ public:
+  CudaDimProto& proto;
+
+  static const uint64_t defaultDim = 1;
+};
+
+/// "Materialized" CudaDimView.
+///
+/// When constructed from values, ignores trailing defaultDim, e.g.,
+///
+///   CudaDim(42, defaultDim);
+///
+/// will only set x, but
+///
+///   CudaDim(42, defaultDim, 32);
+///
+/// will set x, y and z.
+class CudaDim {
+ public:
+  CudaDim() : ownedProto_(), view(ownedProto_) {}
+  CudaDim(const CudaDim& cudaDim)
+      : ownedProto_(cudaDim.ownedProto_), view(ownedProto_) {}
+  CudaDim(const CudaDimProto& proto) : ownedProto_(proto), view(ownedProto_) {}
+  CudaDim(const CudaDimView& view)
+      : ownedProto_(view.proto), view(ownedProto_) {}
+  inline CudaDim(std::initializer_list<uint64_t> il);
+  inline CudaDim(std::vector<uint64_t> il);
+  inline CudaDim(
+      uint64_t x,
+      uint64_t y = CudaDimView::defaultDim,
+      uint64_t z = CudaDimView::defaultDim);
+
+  inline bool operator==(const CudaDim& other) const {
+    return view == other.view;
+  }
+
+ private:
+  CudaDimProto ownedProto_;
+
+ public:
+  CudaDimView view;
+};
+
+/// Specializing CudaDim to differentiate between Block and Grid sizes.
+class Block : public CudaDim {
+ public:
+  Block() = default;
+  Block(const CudaDimView& view) : CudaDim(view.proto) {}
+  Block(const CudaDimProto& proto) : CudaDim(proto) {}
+  Block(std::initializer_list<uint64_t> il) : CudaDim(il) {}
+  Block(std::vector<uint64_t> il) : CudaDim(il) {}
+
+  using CudaDim::operator=;
+};
+
+/// Specializing CudaDim to differentiate between Block and Grid sizes.
+class Grid : public CudaDim {
+ public:
+  Grid() = default;
+  Grid(const CudaDimView& view) : CudaDim(view.proto) {}
+  Grid(const CudaDimProto& proto) : CudaDim(proto) {}
+  Grid(std::initializer_list<uint64_t> il) : CudaDim(il) {}
+  Grid(std::vector<uint64_t> il) : CudaDim(il) {}
+
+  using CudaDim::operator=;
+};
+
+class CudaMappingOptions {
+ private:
+  inline CudaMappingOptions();
+  static CudaMappingOptions makeUnmappedCudaMappingOptions();
+
+ public:
+  /// Construct a deep copy of the options.
+  inline CudaMappingOptions(const CudaMappingOptions& options);
+  inline explicit CudaMappingOptions(const CudaMappingOptionsProto& buf);
+  inline CudaMappingOptions& operator=(const CudaMappingOptions& options);
+
+  /// Compare with another message.
+  inline bool operator==(const CudaMappingOptions& options) const;
+  inline bool operator!=(const CudaMappingOptions& options) const;
+
+  /// Construct from a serialized protocol buffer message.
+  inline explicit CudaMappingOptions(const std::string& str);
+
+  std::string toProtobufSerializedString() const {
+    return ownedProto_.SerializeAsString();
```

Comment:
so even though classes are separated, underlying protobuf representations stay the same. Just wondering, did you consider protobuf separate as well? So different backends don't pollute the same protobuf with options that make sense only in their context.

Reply:
Full separation requires copy-pasting the contents of the protos and then require reimplementing the mutation functions for all classes, so I am not sure it buys us much?

Reply:
How about this?

message MappingOptionsProto {
  optional CudaMappingOptionsProto cuda_options,
  optional CPUMappingOptionsProto cpu_option,
  <the  rest of common options goes here>
}

the in the code we won't accept a message where both cpu and cuda mapping options are set. This emulates union support which is not available in protobuf (is available in thrift).